cpmpy/transformations/decompose_global.py

```diff
@@ -1,107 +1,134 @@
+                if (nested and expr.name not in supported_nested) or (not nested and expr.name not in supported):
+                    aux = intvar(*expr.get_bounds())
+                    newlist.append(aux)
+                    l1, l2 = expr.decompose_comparison("==", aux)
+                    newlist.extend(l1 + l2) # all constraints should be added toplevel, node replaced by aux var
+                else:
+                    newlist.append(expr)
+
+        elif isinstance(expr, Comparison):
+            # Tricky case, if we just recurse into arguments, we risk creating unnecessary auxiliary variables
+            #  e.g., min(x,y,z) == a would become `min(x,y,z).decompose_comparison('==',aux) + [aux == a]`
+            #   while more optimally, its just `min(x,y,z).decompose_comparison('==', a)`
+            #    so have to be careful here and operate from 'one level above'
+
+            lhs, rhs = expr.args
+            if hasattr(lhs, "decompose_comparison"):
+                lhs = copy.copy(lhs)
+                lhs.args = decompose_in_tree(lhs.args, supported, supported_nested, _toplevel, nested=True)
+                lhs_supported = (nested and lhs.name in supported_nested) or (not nested and lhs.name in supported)
+                if not lhs_supported:
+                    newexpr, _toplevel[len(_toplevel):] = lhs.decompose_comparison(expr.name, rhs)
+                    expr = decompose_in_tree(newexpr, supported, supported_nested, _toplevel, nested=nested)  # handle rhs
+                    newlist.append(all(expr))  # this can create toplevel-and
+                    continue
+            else:
+                lhs = decompose_in_tree([lhs], supported, supported_nested, _toplevel, nested=True)
+                lhs = lhs[0]
+
+            if hasattr(rhs, "decompose_comparison"):
+                rhs = copy.copy(rhs)
+                rhs.args = decompose_in_tree(rhs.args, supported, supported_nested, _toplevel, nested=True)
+                rhs_supported = (nested and rhs.name in supported_nested) or (not nested and rhs.name in supported)
+                if not rhs_supported:
+                    # we now lhs is supported, so just flip comparison
+                    newexpr = eval_comparison(flipmap[expr.name], rhs, lhs)
+                    newexpr = decompose_in_tree([newexpr], supported, supported_nested, _toplevel, nested=nested)
+                    newlist.append(all(newexpr))  # can create toplevel and
+                    continue
+            else:
+                rhs = decompose_in_tree([rhs], supported, supported_nested, _toplevel, nested=True)
+                rhs = rhs[0]
+
+            # recreate original comparison
+            newlist.append(eval_comparison(expr.name, lhs, rhs))
+            continue
+
+        else:  # constants, variables, direct constraints
+            newlist.append(expr)
+
+    if nested is True:
+        return newlist
+    elif len(_toplevel):
```

Comment:
I prefer an explicit condition e..g len(topl)>0
(and then to have the ==0 case first ; )

note that in the else you do toplevel_list(newlist) while in the other you do newlist + ... without topleveling it first.